apps/opt.c

```diff
@@ -656,22 +659,54 @@ int opt_next(void)
             /* Just a string. */
             break;
         case '/':
-            if (app_isdir(arg) >= 0)
+            if (app_isdir(arg) > 0)
                 break;
             BIO_printf(bio_err, "%s: Not a directory: %s\n", prog, arg);
             return -1;
         case '<':
             /* Input file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) >= 0)
+            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) == 0)
                 break;
             BIO_printf(bio_err,
                        "%s: Cannot open input file %s, %s\n",
                        prog, arg, strerror(errno));
             return -1;
         case '>':
             /* Output file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
+            l = strlen(arg);
+            if (arg[l - 1] == '/' || app_isdir(arg) > 0) {
```

Comment:
This is exact where the problem is. `app_access(arg, W_OK) >= 0 || errno == ENOENT` can *not* judge if the `arg` is writable, e.g. `/tmp/non-exist-dir/intended-output-filename`, will also be true with this statement (and also not a directory with `app_isdir`)

One possible way is to get the 'parent' directory before calling
```
if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
```
And call `app_access` with the parent directory path. And this definitely requires something like `dirname` involved (or `app_dirname` implemented in a platform-indepent way, such as using `strrchr` in POSIX platform and using other methods to parse Windows and VMS path patterns)